compiler/src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -2871,14 +2871,14 @@ object Parsers {
       if (isIdent(nme.raw.BAR)) { in.nextToken(); pattern1(location) :: patternAlts(location) }
       else Nil
 
-    /**  Pattern1     ::= PatVar Ascription
-     *                  | [‘-’] integerLiteral Ascription
-     *                  | [‘-’] floatingPointLiteral Ascription
-     *                  | Pattern2
+    /**  Pattern1     ::= Pattern2 [Ascription]
      */
     def pattern1(location: Location = Location.InPattern): Tree =
       val p = pattern2()
-      if (isVarPattern(p) || p.isInstanceOf[Number]) && in.isColon then
+      if in.isColon then
+        val isVariableOrNumber = isVarPattern(p) || p.isInstanceOf[Number]
+        if !isVariableOrNumber then
+          warning(em"Only variable and number literal patterns can have type ascriptions")
```

Comment:
We should have a proper test for this. `tests/neg/t5702-neg-bad-and-wild.check` does not count since it is a syntax error in other ways already.

Reply:
I think `tests/neg-custom-args/fatal-warnings/i15893.scala` and `tests/neg-custom-args/fatal-warnings/i10994.scala` are enough as a test.

Reply:
I overlooked i10994.scala. Yes, agreed, that's enough.